docs/miscellaneous.rst

```diff
@@ -88,6 +92,264 @@ by checking if the lowest bit is set: short (not set) and long (set).
 .. note::
   Handling invalidly encoded slots is currently not supported but may be added in the future.
 
+JSON Output
+===========
+
+.. _storage-layout-top-level:
+
+The storage layout of a contract can be requested via the :ref:`standard JSON interface <compiler-api>`.  The output is a JSON object containing two keys,
+``storage`` and ``types``.  The ``storage`` object is an array where each
+element has the following form:
+
+
+.. code::
+
+
+    {
+        "astId":"2",
+        "contract":"A",
+        "label":"x",
+        "offset":"0",
+        "slot":"0",
+        "type":"t_uint256"
+    }
+
+where the example above is the storage layout of ``contract A { uint x; }``
+and
+
+- ``astId`` is the id of the AST node of the state variable
+- ``contract`` is the name of the contract
+- ``label`` is the name of the state variable
+- ``offset`` is the offset within the storage slot according to the encoding
+- ``slot`` is the storage slot where the state variable resides
+- ``type`` is an identifier used as key to the variable's type information
+
+The given ``type``, in this case ``t_uint256`` represents an element in
+``types``, which has the form:
+
+
+.. code::
+
+    {
+        "encoding":"inplace",
+        "label":"uint256",
+        "numberOfBytes":"32",
+    }
+
+where
+
+- ``encoding`` how the data is encoded in storage, where the possible values are:
+
+  - ``inplace``: data is laid out contiguously in storage (see :ref:`above <storage-inplace-encoding>`).
+  - ``mapping``: Keccak-256 hash-based method (see :ref:`above <storage-hashed-encoding>`).
+  - ``dynamic_array``: Keccak-256 hash-based method (see :ref:`above <storage-hashed-encoding>`).
+  - ``string``: single slot or Keccak-256 hash-based depending on the data size (see :ref:`above <bytes-and-string>`).
```

Comment:
Historically (and in the code-base) `bytes` is the generic term, but I'm fine with either